src/treq/test/httpbin_server.py

```diff
@@ -0,0 +1,224 @@
+"""
+A local ``httpbin`` server to run integration tests against.
+
+This ensures tests do not depend on `httpbin <https://httpbin.org/>`_.
+"""
+from __future__ import print_function
+import attr
+import argparse
+import datetime
+import json
+import sys
+
+import httpbin
+
+from twisted.internet.defer import Deferred
+from twisted.internet.endpoints import TCP4ServerEndpoint, SSL4ServerEndpoint
+from twisted.internet.task import react
+from twisted.internet.ssl import (Certificate,
+                                  CertificateOptions)
+
+from OpenSSL.crypto import PKey, X509
+
+from twisted.python.threadpool import ThreadPool
+from twisted.web.server import Site
+from twisted.web.wsgi import WSGIResource
+
+from cryptography import x509
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives import hashes
+from cryptography.hazmat.primitives.asymmetric import rsa
+
+from cryptography.x509.oid import NameOID
+from cryptography.hazmat.primitives.serialization import Encoding
+
+
+@attr.s
+class _HTTPBinDescription(object):
+    """
+    Describe an ``httpbin`` process.
+
+    :param host: The host on which the process listens.
+    :type host: :py:class:`str`
+
+    :param port: The port on which the process listens.
+    :type port: :py:class:`int`
+
+    :param cacert: (optional) The PEM-encoded certificate authority's
+        certificate.  The calling process' treq must trust this when
+        running HTTPS tests.
+    :type cacert: :py:class:`str` or :py:class:`None`
+    """
+    host = attr.ib()
+    port = attr.ib()
+    cacert = attr.ib(default=None)
+
+    @classmethod
+    def from_dictionary(cls, dictionary):
+        return cls(**dictionary)
+
+
+def certificatesForAuthorityAndServer(serviceIdentity):
```

Comment:
If this looks familiar, that's because it's liberated from [Twisted](https://github.com/twisted/twisted/blob/trunk/src/twisted/test/test_sslverify.py#L152-L234)